ccl/sqlproxyccl: TestDirectoryConnect failed

ccl/sqlproxyccl.TestDirectoryConnect [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4421878&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4421878&tab=artifacts#/) on master @ [1a89121d67fa02d5c674766ea01d3c78455fd28b](https://github.com/cockroachdb/cockroach/commits/1a89121d67fa02d5c674766ea01d3c78455fd28b):


```
=== RUN   TestDirectoryConnect
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/fe188f163e90a7ffb92e92cbaea22b0c/logTestDirectoryConnect1677882114
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestDirectoryConnect
    proxy_handler_test.go:753: -- test log scope end --
--- FAIL: TestDirectoryConnect (8.02s)
=== RUN   TestDirectoryConnect/drain_connection
    proxy_handler_test.go:727: 
        	Error Trace:	proxy_handler_test.go:1089
        	Error:      	Expected nil, but got: &sqlproxyccl.codeError{code:4, err:(*withstack.withStack)(0xc003f81d40)}
        	Test:       	TestDirectoryConnect/drain_connection
    --- FAIL: TestDirectoryConnect/drain_connection (2.01s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
/cc @cockroachdb/sql-experience @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDirectoryConnect.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Trying to reproduce the issue through `make` yields no results:

```
[jay@archome] cockroach (master)$ go test -test.v -exec 'stress -p=1' -v ./pkg/ccl/sqlproxyccl/ -run TestDirectoryConnect -timeout 0
0 runs so far, 0 failures, over 5s
1 runs so far, 0 failures, over 10s
1 runs so far, 0 failures, over 15s
2 runs so far, 0 failures, over 20s
3 runs so far, 0 failures, over 25s
3 runs so far, 0 failures, over 30s
4 runs so far, 0 failures, over 35s
4 runs so far, 0 failures, over 40s
5 runs so far, 0 failures, over 45s
6 runs so far, 0 failures, over 50s
6 runs so far, 0 failures, over 55s
7 runs so far, 0 failures, over 1m0s
7 runs so far, 0 failures, over 1m5s
8 runs so far, 0 failures, over 1m10s
9 runs so far, 0 failures, over 1m15s
9 runs so far, 0 failures, over 1m20s
10 runs so far, 0 failures, over 1m25s
10 runs so far, 0 failures, over 1m30s
11 runs so far, 0 failures, over 1m35s
12 runs so far, 0 failures, over 1m40s
12 runs so far, 0 failures, over 1m45s
13 runs so far, 0 failures, over 1m50s
14 runs so far, 0 failures, over 1m55s
14 runs so far, 0 failures, over 2m0s
15 runs so far, 0 failures, over 2m5s
15 runs so far, 0 failures, over 2m10s
16 runs so far, 0 failures, over 2m15s
17 runs so far, 0 failures, over 2m20s
17 runs so far, 0 failures, over 2m25s
18 runs so far, 0 failures, over 2m30s
18 runs so far, 0 failures, over 2m35s
19 runs so far, 0 failures, over 2m40s
20 runs so far, 0 failures, over 2m45s
20 runs so far, 0 failures, over 2m50s
21 runs so far, 0 failures, over 2m55s
21 runs so far, 0 failures, over 3m0s
22 runs so far, 0 failures, over 3m5s
23 runs so far, 0 failures, over 3m10s
23 runs so far, 0 failures, over 3m15s
24 runs so far, 0 failures, over 3m20s
25 runs so far, 0 failures, over 3m25s
25 runs so far, 0 failures, over 3m30s
26 runs so far, 0 failures, over 3m35s
26 runs so far, 0 failures, over 3m40s
27 runs so far, 0 failures, over 3m45s
28 runs so far, 0 failures, over 3m50s
28 runs so far, 0 failures, over 3m55s
29 runs so far, 0 failures, over 4m0s
29 runs so far, 0 failures, over 4m5s
30 runs so far, 0 failures, over 4m10s
31 runs so far, 0 failures, over 4m15s
31 runs so far, 0 failures, over 4m20s
32 runs so far, 0 failures, over 4m25s
33 runs so far, 0 failures, over 4m30s
33 runs so far, 0 failures, over 4m35s
34 runs so far, 0 failures, over 4m40s
34 runs so far, 0 failures, over 4m45s
35 runs so far, 0 failures, over 4m50s
36 runs so far, 0 failures, over 4m55s
36 runs so far, 0 failures, over 5m0s
37 runs so far, 0 failures, over 5m5s
37 runs so far, 0 failures, over 5m10s
38 runs so far, 0 failures, over 5m15s
39 runs so far, 0 failures, over 5m20s
39 runs so far, 0 failures, over 5m25s
40 runs so far, 0 failures, over 5m30s
40 runs so far, 0 failures, over 5m35s
41 runs so far, 0 failures, over 5m40s
42 runs so far, 0 failures, over 5m45s
42 runs so far, 0 failures, over 5m50s
43 runs so far, 0 failures, over 5m55s
43 runs so far, 0 failures, over 6m0s
44 runs so far, 0 failures, over 6m5s
45 runs so far, 0 failures, over 6m10s
45 runs so far, 0 failures, over 6m15s
46 runs so far, 0 failures, over 6m20s
47 runs so far, 0 failures, over 6m25s
47 runs so far, 0 failures, over 6m30s
48 runs so far, 0 failures, over 6m35s
48 runs so far, 0 failures, over 6m40s
49 runs so far, 0 failures, over 6m45s
50 runs so far, 0 failures, over 6m50s
50 runs so far, 0 failures, over 6m55s
51 runs so far, 0 failures, over 7m0s
51 runs so far, 0 failures, over 7m5s
52 runs so far, 0 failures, over 7m10s
53 runs so far, 0 failures, over 7m15s
53 runs so far, 0 failures, over 7m20s
54 runs so far, 0 failures, over 7m25s
55 runs so far, 0 failures, over 7m30s
55 runs so far, 0 failures, over 7m35s
56 runs so far, 0 failures, over 7m40s
^C56 runs completed, 0 failures, over 7m43s
context canceled
FAIL
FAIL	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl	463.322s
```

Running it on bazel with a high concurrency count leads to an issue that looks like the following:
```
./dev test ./pkg/ccl/sqlproxyccl --cpus 30 -f=TestDirectoryConnect -v --stress --test-args='-test.timeout=10s'
$ bazel test --local_cpu_resources=30 --test_sharding_strategy=disabled pkg/ccl/sqlproxyccl:all --test_env=GOTRACEBACK=all --test_timeout=86400 --run_under '@com_github_cockroachdb_stress//:stress -bazel -shardable-artifacts '"'"'XML_OUTPUT_FILE=/home/jay/workspace/src/github.com/cockroachdb/cockroach/bin/dev-versions/dev.17 merge-test-xmls'"'"' -p=30' --test_filter=TestDirectoryConnect --test_sharding_strategy=disabled --test_arg -test.v --test_arg -test.outputdir=/home/jay/workspace/src/github.com/cockroachdb/cockroach --sandbox_writable_path=/home/jay/workspace/src/github.com/cockroachdb/cockroach --test_arg -test.timeout=10s --test_output streamed
Starting local Bazel server and connecting to it...
INFO: Invocation ID: 5f9698ce-d22f-4cd5-9718-5a70059f139d
WARNING: Streamed test output requested. All tests will be run locally, without sharding, one at a time
INFO: Analyzed 3 targets (1251 packages loaded, 20818 targets configured).
INFO: Found 2 targets and 1 test target...
WARNING: Remote Cache: Connection refused: /127.0.0.1:9867
0 runs so far, 0 failures, over 5s
30 runs so far, 0 failures, over 10s
30 runs so far, 0 failures, over 15s
60 runs so far, 0 failures, over 20s
76 runs so far, 0 failures, over 25s
90 runs so far, 0 failures, over 30s
120 runs so far, 0 failures, over 35s
120 runs so far, 0 failures, over 40s
150 runs so far, 0 failures, over 45s
179 runs so far, 0 failures, over 50s
180 runs so far, 0 failures, over 55s
210 runs so far, 0 failures, over 1m0s
211 runs so far, 0 failures, over 1m5s
240 runs so far, 0 failures, over 1m10s
270 runs so far, 0 failures, over 1m15s
270 runs so far, 0 failures, over 1m20s
300 runs so far, 0 failures, over 1m25s
310 runs so far, 0 failures, over 1m30s
330 runs so far, 0 failures, over 1m35s
360 runs so far, 0 failures, over 1m40s
360 runs so far, 0 failures, over 1m45s
390 runs so far, 0 failures, over 1m50s
412 runs so far, 0 failures, over 1m55s
420 runs so far, 0 failures, over 2m0s

initialized metamorphic constant "span-reuse-rate" with value 79
initialized metamorphic constant "coldata-batch-size" with value 3989
initialized metamorphic constant "addsst-rewrite-concurrency" with value 7
initialized metamorphic constant "kv.refresh_range.time_bound_iterators.enabled" with value false
initialized metamorphic constant "raft-log-truncation-clearrange-threshold" with value 129597
initialized metamorphic constant "kv.rangefeed.catchup_scan_iterator_optimization.enabled" with value false
initialized metamorphic constant "kv-batch-size" with value 1
initialized metamorphic constant "datum-row-converter-batch-size" with value 1
initialized metamorphic constant "row-container-rows-per-chunk-shift" with value 1
initialized metamorphic constant "inverted-joiner-batch-size" with value 1
initialized metamorphic constant "zig-zag-joiner-batch-size" with value 1
initialized metamorphic constant "spilling-queue-initial-len" with value 15
initialized metamorphic constant "merge-joiner-groups-buffer" with value 9
initialized metamorphic constant "ColIndexJoin-batch-size" with value 3187065
initialized metamorphic constant "max-batch-byte-size" with value 29862101
initialized metamorphic constant "async-IE-result-channel-buffer-size" with value 25
I220223 01:49:29.110408 1 (gostd) rand.go:147  [-] 1  random seed: 7902158050586388781
=== RUN   TestDirectoryConnect
    test_log_scope.go:79: test logs captured to: /home/jay/.cache/bazel/_bazel_jay/ad30d6d5ac3ec70478dfe5900736ea48/sandbox/linux-sandbox/1/execroot/cockroach/_tmp/a455c59caed0cf9d0a0298ad88bfe7a2/logTestDirectoryConnect1906532003
    test_log_scope.go:80: use -show-logs to present logs inline
=== RUN   TestDirectoryConnect/fallback_when_tenant_not_found
=== RUN   TestDirectoryConnect/fail_to_connect_to_backend
=== RUN   TestDirectoryConnect/successful_connection
=== RUN   TestDirectoryConnect/successful_connection_after_restart
=== RUN   TestDirectoryConnect/drain_connection
    proxy_handler_test.go:727: 
        	Error Trace:	proxy_handler_test.go:1090
        	Error:      	Received unexpected error:
        	            	while receiving startup message
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.newErrorf
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/error.go:100
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.glob..func4
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/frontend_admitter.go:35
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*proxyHandler).handle
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/proxy_handler.go:206
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*Server).Serve.func2
        	            	  | 	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/server.go:185
        	            	  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:494
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1581
        	            	Wraps: (2) while receiving startup message
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestDirectoryConnect/drain_connection
=== CONT  TestDirectoryConnect
    proxy_handler_test.go:753: -- test log scope end --
test logs left over in: /home/jay/.cache/bazel/_bazel_jay/ad30d6d5ac3ec70478dfe5900736ea48/sandbox/linux-sandbox/1/execroot/cockroach/_tmp/a455c59caed0cf9d0a0298ad88bfe7a2/logTestDirectoryConnect1906532003
--- FAIL: TestDirectoryConnect (8.13s)
    --- PASS: TestDirectoryConnect/fallback_when_tenant_not_found (0.00s)
    --- PASS: TestDirectoryConnect/fail_to_connect_to_backend (0.23s)
    --- PASS: TestDirectoryConnect/successful_connection (0.00s)
    --- PASS: TestDirectoryConnect/successful_connection_after_restart (2.57s)
    --- FAIL: TestDirectoryConnect/drain_connection (2.01s)
FAIL
I220223 01:49:37.250692 1 (gostd) testmain.go:154  [-] 1  Test //pkg/ccl/sqlproxyccl:sqlproxyccl_test exited with error code 1


ERROR: exit status 1

446 runs completed, 1 failures, over 2m4s
context canceled
FAIL
FAIL: //pkg/ccl/sqlproxyccl:sqlproxyccl_test (see /home/jay/.cache/bazel/_bazel_jay/ad30d6d5ac3ec70478dfe5900736ea48/execroot/cockroach/bazel-out/k8-fastbuild/testlogs/pkg/ccl/sqlproxyccl/sqlproxyccl_test/test.log)
INFO: Elapsed time: 129.758s, Critical Path: 125.63s
INFO: 2 processes: 1 internal, 1 linux-sandbox.
INFO: Build completed, 1 test FAILED, 2 total actions
//pkg/ccl/sqlproxyccl:sqlproxyccl_test                                   FAILED in 125.1s
  /home/jay/.cache/bazel/_bazel_jay/ad30d6d5ac3ec70478dfe5900736ea48/execroot/cockroach/bazel-out/k8-fastbuild/testlogs/pkg/ccl/sqlproxyccl/sqlproxyccl_test/test.log

INFO: Build completed, 1 test FAILED, 2 total actions
ERROR: exit status 3

```

When I printed out the error message, I see:

```
while receiving startup message: invalid length of startup packet: 1347569948
```

This particular test exercises the idle monitor. For some reason when the idle monitor gets triggered, clients attempt to send messages with an invalid startup packet. Looking at the code, when that case happens, we close both the client's and server's connections. I'm not sure why we're getting this issue yet, and I'll need to dig deeper. I'll poke more tomorrow.

That said, given that we're going to remove the idle monitor, and it's currently not being used in production, I'm leaning towards skipping the test under stress.

cc: @JeffSwenson @andy-kimball

--- a/pkg/ccl/sqlproxyccl/BUILD.bazel
+++ b/pkg/ccl/sqlproxyccl/BUILD.bazel
@@ -101,7 +101,6 @@ go_test(
         "//pkg/testutils",
         "//pkg/testutils/datapathutils",
         "//pkg/testutils/serverutils",
-        "//pkg/testutils/skip",
         "//pkg/testutils/sqlutils",
         "//pkg/testutils/testcluster",
         "//pkg/util/leaktest",

--- a/pkg/ccl/sqlproxyccl/proxy_handler_test.go
+++ b/pkg/ccl/sqlproxyccl/proxy_handler_test.go
@@ -42,7 +42,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/datapathutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -57,8 +56,6 @@ import (
 	proxyproto "github.com/pires/go-proxyproto"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
-	"google.golang.org/grpc/codes"
-	"google.golang.org/grpc/status"
 	"gopkg.in/yaml.v3"
 )
 
@@ -966,51 +963,41 @@ func TestDenylistUpdate(t *testing.T) {
 
 func TestDirectoryConnect(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-
-	skip.WithIssue(t, 76839, "flaky test")
-	// TODO(jaylim-crl): This is a potential port reuse issue, so skip this
-	// under stress. See linked GitHub issue.
-	skip.UnderStress(t, "https://github.com/cockroachdb/cockroach/issues/76839")
-	skip.UnderDeadlockWithIssue(t, 71365)
 	defer log.Scope(t).Close(t)
 
 	ctx := context.Background()
 	te := newTester()
 	defer te.Close()
 
-	srv, _, _ := serverutils.StartServer(t,
-		base.TestServerArgs{
-			Insecure: true,
-			// Need to disable the test tenant here because it appears as though
-			// we're not able to establish the necessary connections from within
-			// it. More investigation required (tracked with #76378).
-			DefaultTestTenant: base.TestTenantDisabled,
-		},
-	)
-	srv.(*server.TestServer).PGPreServer().TestingSetTrustClientProvidedRemoteAddr(true)
-	defer srv.Stopper().Stop(ctx)
+	// Start KV server.
+	params, _ := tests.CreateTestServerParams()
+	s, _, _ := serverutils.StartServer(t, params)
+	defer s.Stopper().Stop(ctx)
 
-	// Create tenant 28.
-	sqlConn := srv.InternalExecutor().(*sql.InternalExecutor)
-	_, err := sqlConn.Exec(ctx, "", nil, "SELECT crdb_internal.create_tenant(28)")
-	require.NoError(t, err)
+	// Start a SQL pod for the test tenant, and register it with the directory
+	// server. Use a custom stopper so we can stop the tenant later.
+	tenantStopper := stop.NewStopper()
+	defer tenantStopper.Stop(ctx)
+	tenantID := serverutils.TestTenantID()
+	tenants := startTestTenantPodsWithStopper(ctx, t, s, tenantID, 1, base.TestingKnobs{}, tenantStopper)
+	tds := tenantdirsvr.NewTestStaticDirectoryServer(s.Stopper(), nil /* timeSource */)
+	tds.CreateTenant(tenantID, "tenant-cluster")
+	tds.AddPod(tenantID, &tenant.Pod{
+		TenantID:       tenantID.ToUint64(),
+		Addr:           tenants[0].SQLAddr(),
+		State:          tenant.RUNNING,
+		StateTimestamp: timeutil.Now(),
+	})
+	require.NoError(t, tds.Start(ctx))
 
-	// New test directory server.
-	tds1, tdsAddr := newDirectoryServer(ctx, t, srv, &net.TCPAddr{})
+	// Start the proxy server using the static directory server.
+	opts := &ProxyOptions{SkipVerify: true}
+	opts.testingKnobs.directoryServer = tds
+	_, addr, _ := newSecureProxyServer(ctx, t, s.Stopper(), opts)
+	connectionString := fmt.Sprintf("postgres://testuser:hunter2@%s/?sslmode=require&options=--cluster=tenant-cluster-%s", addr, tenantID)
 
-	// New proxy server using the directory. Define both the directory and the
-	// routing rule so that fallback to the routing rule can be tested.
-	opts := &ProxyOptions{
-		RoutingRule:   srv.ServingSQLAddr(),
-		DirectoryAddr: tdsAddr.String(),
-		Insecure:      true,
-	}
-	_, addr, _ := newProxyServer(ctx, t, srv.Stopper(), opts)
-
-	t.Run("fallback when tenant not found", func(t *testing.T) {
-		url := fmt.Sprintf(
-			"postgres://root:admin@%s/?sslmode=disable&options=--cluster=tenant-cluster-%d",
-			addr, notFoundTenantID)
+	t.Run("tenant not found", func(t *testing.T) {
+		url := fmt.Sprintf("postgres://testuser:hunter2@%s/?sslmode=require&options=--cluster=tenant-cluster-%d", addr, notFoundTenantID)
 		_ = te.TestConnectErr(ctx, t, url, codeParamsRoutingFailed, "cluster tenant-cluster-99 not found")
 	})
 
@@ -1030,50 +1017,51 @@ func TestDirectoryConnect(t *testing.T) {
 		// Ensure that Directory.ReportFailure is being called correctly.
 		countReports := 0
 		defer testutils.TestingHook(&reportFailureToDirectoryCache, func(
-			ctx context.Context, tenantID roachpb.TenantID, addr string, directoryCache tenant.DirectoryCache,
+			ctx context.Context, tenID roachpb.TenantID, addr string, directoryCache tenant.DirectoryCache,
 		) error {
-			require.Equal(t, roachpb.MustMakeTenantID(28), tenantID)
-			pods, err := directoryCache.TryLookupTenantPods(ctx, tenantID)
+			require.Equal(t, tenantID, tenID)
+			pods, err := directoryCache.TryLookupTenantPods(ctx, tenID)
 			require.NoError(t, err)
 			require.Len(t, pods, 1)
 			require.Equal(t, pods[0].Addr, addr)
 
 			countReports++
-			err = directoryCache.ReportFailure(ctx, tenantID, addr)
+			err = directoryCache.ReportFailure(ctx, tenID, addr)
 			require.NoError(t, err)
 			return err
 		})()
 
-		url := fmt.Sprintf("postgres://root:admin@%s/?sslmode=disable&options=--cluster=tenant-cluster-28", addr)
-		_ = te.TestConnectErr(ctx, t, url, codeBackendDisconnected, "backend disconnected")
+		_ = te.TestConnectErr(ctx, t, connectionString, codeBackendDisconnected, "backend disconnected")
 		require.Equal(t, 3, countFailures)
 		require.Equal(t, 2, countReports)
 	})
 
 	t.Run("successful connection", func(t *testing.T) {
-		url := fmt.Sprintf("postgres://root:admin@%s/?sslmode=disable&options=--cluster=tenant-cluster-28", addr)
-		te.TestConnect(ctx, t, url, func(conn *pgx.Conn) {
+		te.TestConnect(ctx, t, connectionString, func(conn *pgx.Conn) {
 			require.NoError(t, runTestQuery(ctx, conn))
 		})
 	})
 
-	// Stop the directory server and the tenant SQL process started earlier. This
-	// tests whether the proxy can recover when the directory server and a SQL
-	// tenant pod restart.
-	tds1.Stopper().Stop(ctx)
+	// Stop the directory server and the tenant SQL process started earlier.
+	// This tests whether the proxy can recover when the directory server and
+	// SQL pod restarts.
+	tds.Stop(ctx)
+	tenantStopper.Stop(ctx)
 
-	// Pass the same tdsAddr used to start up the directory server previously,
-	// since it's not allowed to jump to a different address.
-	tds2, _ := newDirectoryServer(ctx, t, srv, tdsAddr)
-	defer tds2.Stopper().Stop(ctx)
+	// Drain old pod and add a new one before starting the directory server.
+	tds.DrainPod(tenantID, tenants[0].SQLAddr())
+	tenants = startTestTenantPods(ctx, t, s, tenantID, 1, base.TestingKnobs{})
+	tds.AddPod(tenantID, &tenant.Pod{
+		TenantID:       tenantID.ToUint64(),
+		Addr:           tenants[0].SQLAddr(),
+		State:          tenant.RUNNING,
+		StateTimestamp: timeutil.Now(),
+	})
+	require.NoError(t, tds.Start(ctx))
 
 	t.Run("successful connection after restart", func(t *testing.T) {
-		// Try to connect through the proxy again. This may take several tries
-		// in order to clear the proxy directory of the old SQL tenant process
-		// address and replace with the new.
 		require.Eventually(t, func() bool {
-			url := fmt.Sprintf("postgres://root:admin@%s/?sslmode=disable&options=--cluster=tenant-cluster-28", addr)
-			conn, err := pgx.Connect(ctx, url)
+			conn, err := pgx.Connect(ctx, connectionString)
 			if err != nil {
 				return false
 			}
@@ -2602,49 +2590,6 @@ func runTestQuery(ctx context.Context, conn *pgx.Conn) error {
 	return nil
 }
 
-func newDirectoryServer(
-	ctx context.Context, t *testing.T, srv serverutils.TestServerInterface, addr *net.TCPAddr,
-) (*tenantdirsvr.TestDirectoryServer, *net.TCPAddr) {
-	// Start listening on port that the tenant directory server will use.
-	var listener *net.TCPListener
-	require.Eventually(t, func() bool {
-		var err error
-		listener, err = net.ListenTCP("tcp", addr)
-		return err == nil
-	}, 30*time.Second, time.Second)
-
-	// Create the tenant directory server.
-	tdsStopper := stop.NewStopper()
-	tds, err := tenantdirsvr.New(tdsStopper)
-	require.NoError(t, err)
-
-	// Override the tenant starter function to start a new tenant process using
-	// the TestServerInterface.
-	tds.TenantStarterFunc = func(ctx context.Context, tenantID uint64) (*tenantdirsvr.Process, error) {
-		// Recognize special tenant ID that triggers an error.
-		if tenantID == notFoundTenantID {
-			return nil, status.Error(codes.NotFound, "tenant not found")
-		}
-
-		tenantStopper := tenantdirsvr.NewSubStopper(tdsStopper)
-		ten, err := srv.StartTenant(ctx, base.TestTenantArgs{
-			TenantID:      roachpb.MustMakeTenantID(tenantID),
-			ForceInsecure: true,
-			Stopper:       tenantStopper,
-		})
-		require.NoError(t, err)
-		sqlAddr, err := net.ResolveTCPAddr("tcp", ten.SQLAddr())
-		require.NoError(t, err)
-		ten.(*server.TestTenant).PGPreServer().TestingSetTrustClientProvidedRemoteAddr(true)
-		return &tenantdirsvr.Process{SQL: sqlAddr, Stopper: tenantStopper}, nil
-	}
-
-	// Start serving on a background goroutine.
-	go func() { require.NoError(t, tds.Serve(listener)) }()
-
-	return tds, listener.Addr().(*net.TCPAddr)
-}
-
 // mustGetTestSimpleDirectoryServer returns the underlying simple directory
 // server. This can only be used with a routing rule.
 func mustGetTestSimpleDirectoryServer(
@@ -2687,18 +2632,33 @@ func startTestTenantPods(
 	count int,
 	knobs base.TestingKnobs,
 ) []serverutils.TestTenantInterface {
+	return startTestTenantPodsWithStopper(ctx, t, ts, tenantID, count, knobs, nil)
+}
+
+// startTestTenantPodsWithStopper is similar to startTestTenantPods, but allows
+// a custom stopper
+func startTestTenantPodsWithStopper(
+	ctx context.Context,
+	t *testing.T,
+	ts serverutils.TestServerInterface,
+	tenantID roachpb.TenantID,
+	count int,
+	knobs base.TestingKnobs,
+	stopper *stop.Stopper,
+) []serverutils.TestTenantInterface {
 	t.Helper()
 
 	var tenants []serverutils.TestTenantInterface
 	for i := 0; i < count; i++ {
 		params := tests.CreateTestTenantParams(tenantID)
 		params.TestingKnobs = knobs
+		params.Stopper = stopper
 		tenant, tenantDB := serverutils.StartTenant(t, ts, params)
 		tenant.(*server.TestTenant).PGPreServer().TestingSetTrustClientProvidedRemoteAddr(true)
 
 		// Create a test user. We only need to do it once.
 		if i == 0 {
-			_, err := tenantDB.Exec("CREATE USER testuser WITH PASSWORD 'hunter2'")
+			_, err := tenantDB.Exec("CREATE USER IF NOT EXISTS testuser WITH PASSWORD 'hunter2'")
 			require.NoError(t, err)
 			_, err = tenantDB.Exec("GRANT admin TO testuser")
 			require.NoError(t, err)

